xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -672,14 +1048,268 @@ class CudnnNormRewriterVisitor : public DfsHloRewriteVisitor {
+      // broadcasted dimensions.
+      float actual_r_nelems = scalar->literal().GetAsDouble({}).value();
+      int64_t nelems = 1;
+      for (int i = 0; i < broadcast->shape().dimensions_size(); ++i) {
+        if (!c_linear_search(broadcast->dimensions(), i)) {
+          nelems *= broadcast->shape().dimensions()[i];
+        }
+      }
+      // The absolute of the difference between the actual scaling factor and
+      // the reference value must not exceed a prescribed threshold.
+      float r_nelems = 1. / static_cast<float>(nelems);
+      float numerical_epsilon = std::numeric_limits<bfloat16>::epsilon();
+      if (!(abs(actual_r_nelems - r_nelems) <
+            ((actual_r_nelems + r_nelems) * numerical_epsilon))) {
+        VLOG(1)
+            << "Layer norm backward broadcast operand outside expected range.";
+        return OkStatus();
+      }
+
+      // Identify Dscale = AddReduce(XCenter * NormFactor * DY)
+      // through a top-down search starting from XCenter.
+      auto find_dscale = [x_center, fused_norm_factor,
+                          dy]() -> HloInstruction* {
+        for (const HloInstruction* multiply_norm_factor : x_center->users()) {
+          multiply_norm_factor =
+              SkipUnaryOps(multiply_norm_factor, /*top-down=*/true);
+          // Verify that the user of XCenter is a multiply with a single user.
+          if (!multiply_norm_factor ||
+              multiply_norm_factor->opcode() != HloOpcode::kMultiply ||
+              multiply_norm_factor->user_count() != 1) {
+            continue;
+          }
+          // Verify that one of the factors is a broadcast operating on the
+          // fused norm factor.
+          bool bcast_operand = multiply_norm_factor->operand(0)->opcode() !=
+                               HloOpcode::kBroadcast;
+          if (multiply_norm_factor->operand(bcast_operand)->opcode() !=
+                  HloOpcode::kBroadcast ||
+              SkipUnaryOps(
+                  SkipUnaryOps(multiply_norm_factor->operand(bcast_operand))
+                      ->operand(0))
+                      ->unique_id() != fused_norm_factor->unique_id()) {
+            continue;
+          }
+          const HloInstruction* multiply_dy =
+              SkipUnaryOps(multiply_norm_factor->users()[0], /*top-down=*/true);
+          // Verify that the user of the multiply is another multiply that is
+          // also a user of DY.
+          if (!multiply_dy || multiply_dy->opcode() != HloOpcode::kMultiply ||
+              !multiply_dy->IsUserOf(dy)) {
+            continue;
+          }
+          // Dscale applies an addition-reduction to multiply_dy.
+          for (HloInstruction* multiply_dy_user : multiply_dy->users()) {
+            if (AppliesAddReduce(multiply_dy_user)) {
+              return multiply_dy_user;
+            }
+          }
+        }
+        return nullptr;
+      };
+      if (!(dscale = find_dscale())) {
```

Comment:
Our pattern matching library does not easily support top-down pattern matching, making the implementation of `find_dscale` complex. Instead, can you initially not fuse dscale, then later in `HandleReduce` check for dscale in a bottom-up fasion? In general, I think it's better if the pattern matching is more incremental.

And similarly for dbias below.